squirtgun-platform-discord/src/main/java/net/lucypoulton/squirtgun/discord/adventure/DiscordComponentSerializer.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package net.lucypoulton.squirtgun.discord.adventure;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.format.TextDecoration;
+import net.kyori.adventure.text.serializer.ComponentSerializer;
+import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
+import net.lucypoulton.squirtgun.util.Pair;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static net.kyori.adventure.text.format.TextDecoration.State;
+
+/**
+ * Serialises Components into Discord-formatted markdown. Discord's text format is
+ * far less rich than Minecraft's, so some information, such as colour and events,
+ * is discarded.
+ *
+ * Deserialisation is not properly implemented currently, and all formatting is ignored.
+ */
+public enum DiscordComponentSerializer implements ComponentSerializer<Component, TextComponent, String> {
+
+    INSTANCE;
+
+    private static final Map<TextDecoration, String> DECORATION_MARKUP = Map.of(
+        TextDecoration.BOLD, "**",
+        TextDecoration.ITALIC, "*",
+        TextDecoration.UNDERLINED, "__",
+        TextDecoration.STRIKETHROUGH, "~~",
+        TextDecoration.OBFUSCATED, "||"
+    );
+
+    @Override
+    public @NotNull TextComponent deserialize(@NotNull String input) {
+        return Component.text(input);
+    }
+
+
+    private String getStringValue(Component in) {
+        if (in instanceof TextComponent) return ((TextComponent) in).content();
+        return PlainTextComponentSerializer.plainText().serialize(in);
+    }
+
+    private Set<TextDecoration> decorateChild(Component parent, Component child) {
+        return parent.mergeStyle(child).decorations().entrySet().stream()
+            .filter(deco -> deco.getValue() == State.TRUE)
+            .map(Map.Entry::getKey)
+            .collect(Collectors.toSet());
+    }
+
+    private List<Pair<String, Set<TextDecoration>>> flatten(Component component, @Nullable Component parent) {
+        if (parent == null) {
+            parent = Component.empty();
+        }
+
+        List<Pair<String, Set<TextDecoration>>> out = new ArrayList<>();
+        out.add(new Pair<>(getStringValue(component), decorateChild(parent, component)));
+
+        for (Component child : component.children()) {
+            out.addAll(flatten(child, component));
+        }
+        return out;
+    }
```

Comment:
This does not consider decorations inherited transitively. It fails on:
```json
{
  "text": "Parent",
  "bold": true,
  "extra": [
    {
      "text": ", child",
      "extra": [
        {
          "text": ", child's child"
        }
      ]
    }
  ]
}
```
This should render as `**Parent, child, child's child**` but instead renders as `**Parent, child**, child's child`.

I suggest using the newly introduced [component iterator API](https://github.com/KyoriPowered/adventure/pull/339/files) to scan the component tree and handle this case as well, but that's a suggestion, do as you will.

Reply:
resolved in 757c631. will look into the iterator